mlir/tools/mlir-miopen-lib/mlir-miopen-lib.h

```diff
@@ -0,0 +1,85 @@
+//===-------------- mlir-miopen-lib.h - C API ---------------------*- C -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM
+// Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_C_MLIRMIOPEN_LIBMLIRMIOPEN_H
+#define MLIR_C_MLIRMIOPEN_LIBMLIRMIOPEN_H
+
+#include <stddef.h>
+
+#define MLIRMIOPEN_VERSION_FLAT 0
+
+enum MlirmiopenStatus {
+  MlirmiopenSuccess = 0,
+  MlirmiopenInvalidParam = -1,
+  MlirmiopenInvalidModule = -2,
+  MlirmiopenBuildFailure = -3
+};
+typedef enum MlirmiopenStatus MlirmiopenStatus;
+
+/*! @brief The MLIR handle used for lowering and code generation
+ */
+typedef void *MlirmiopenHandle;
+
+/*! @brief Create the MLIR handle according to options string
+ *  @param options Command-line options as a string
+ *  @return        MLIR handle
+ */
+extern "C" MlirmiopenHandle mlirmiopenCreateHandle(const char *options);
+
+/*! @brief Lower the MLIR module to c++ code
+ *  @param handle   MLIR handle
+ */
+extern "C" MlirmiopenStatus mlirmiopenLowerCpp(MlirmiopenHandle handle);
+
+/*! @brief Populate Conv2d implicitgemm host code for MIOpen
+ *  @param handle   MLIR handle
+ *  @return         Source string
+ */
+extern "C" const char *mlirmiopenGenIgemmSource(MlirmiopenHandle handle);
```

Comment:
The function returns a pointer to the buffer that is owned by the library. When the library deallocates that buffer?

What about returning status to the caller? Or this function never fails?

Reply:
@atamazov Wait, why is it not multi-threading/process safe by design? As long as client keep track of their allocated handle, I don't see it has an issue in that space.

Reply:
Because each thread/process must have its own handle OR implement some kind of synchronization. But this is not a problem AFAICS because handles are lightweight.

Reply:
@atamazov I think we are on the same page. Yes, if it is run with multi-process/thread in mind, each thread needs to manage its own handle correctly.

Marking all 3 cpp API related comments as resolved